File: calorie_calc.php
<?php
// calorie_calc.php - Калькулятор калорий на PHP (CLI + веб)
// CLI: php calorie_calc.php --gender=male --age=30 --height=175 --weight=75 --activity=moderate --goal=lose

$activityLevels = [
    'sedentary' => 1.2,
    'light' => 1.375,
    'moderate' => 1.55,
    'active' => 1.725,
    'very_active' => 1.9,
];
$activityNames = [
    'sedentary' => 'Сидячий образ жизни',
    'light' => 'Лёгкая активность (1-3 дня/нед)',
    'moderate' => 'Умеренная активность (3-5 дней/нед)',
    'active' => 'Высокая активность (6-7 дней/нед)',
    'very_active' => 'Очень высокая активность (тяжёлая работа/спорт)',
];
$goalNames = [
    'maintain' => 'Поддержание веса',
    'lose' => 'Похудение (дефицит 15%)',
    'gain' => 'Набор массы (профицит 15%)',
];
$genderNames = ['male' => 'Мужской', 'female' => 'Женский'];

function bmrMifflin($gender, $age, $height, $weight) {
    if ($gender == 'male') {
        return 10 * $weight + 6.25 * $height - 5 * $age + 5;
    } else {
        return 10 * $weight + 6.25 * $height - 5 * $age - 161;
    }
}

function bmrHarris($gender, $age, $height, $weight) {
    if ($gender == 'male') {
        return 88.362 + 13.397 * $weight + 4.799 * $height - 5.677 * $age;
    } else {
        return 447.593 + 9.247 * $weight + 3.098 * $height - 4.330 * $age;
    }
}

function calculateTdee($bmr, $activity) {
    global $activityLevels;
    return $bmr * $activityLevels[$activity];
}

function calculateGoalCalories($tdee, $goal) {
    switch ($goal) {
        case 'maintain': return $tdee;
        case 'lose': return $tdee * 0.85;
        case 'gain': return $tdee * 1.15;
        default: return $tdee;
    }
}

function calculateMacros($calories, $goal) {
    $proteinRatio = 0.30; $fatRatio = 0.25; $carbRatio = 0.45;
    if ($goal == 'lose') {
        $proteinRatio = 0.35; $fatRatio = 0.25; $carbRatio = 0.40;
    } elseif ($goal == 'gain') {
        $proteinRatio = 0.30; $fatRatio = 0.20; $carbRatio = 0.50;
    }
    $protein = ($calories * $proteinRatio) / 4;
    $fat = ($calories * $fatRatio) / 9;
    $carbs = ($calories * $carbRatio) / 4;
    return ['protein' => $protein, 'fat' => $fat, 'carbs' => $carbs];
}

function generateReport($params, $bmr, $tdee, $goalCal, $macros) {
    global $activityNames, $goalNames, $genderNames;
    $lines = [];
    $lines[] = str_repeat('=', 50);
    $lines[] = "📋 ОТЧЁТ ПО КАЛОРИЯМ";
    $lines[] = str_repeat('=', 50);
    $lines[] = "Пол: " . $genderNames[$params['gender']];
    $lines[] = "Возраст: " . $params['age'] . " лет";
    $lines[] = "Рост: " . $params['height'] . " см";
    $lines[] = "Вес: " . $params['weight'] . " кг";
    $lines[] = "Уровень активности: " . $activityNames[$params['activity']];
    $lines[] = "Цель: " . $goalNames[$params['goal']];
    $lines[] = str_repeat('-', 50);
    $lines[] = "BMR (базовый метаболизм): " . number_format($bmr, 1) . " ккал/день";
    $lines[] = "TDEE (суточная норма): " . number_format($tdee, 1) . " ккал/день";
    $lines[] = "Рекомендуемая калорийность для цели: " . number_format($goalCal, 1) . " ккал/день";
    $lines[] = str_repeat('-', 50);
    $lines[] = "🍽️ РАСПРЕДЕЛЕНИЕ ПО МАКРОНУТРИЕНТАМ:";
    $lines[] = "Белки: " . number_format($macros['protein'], 1) . " г (" . number_format(($macros['protein']*4/$goalCal)*100, 0) . "%)";
    $lines[] = "Жиры: " . number_format($macros['fat'], 1) . " г (" . number_format(($macros['fat']*9/$goalCal)*100, 0) . "%)";
    $lines[] = "Углеводы: " . number_format($macros['carbs'], 1) . " г (" . number_format(($macros['carbs']*4/$goalCal)*100, 0) . "%)";
    $lines[] = str_repeat('=', 50);
    return implode("\n", $lines);
}

// ========== CLI ==========
if (php_sapi_name() === 'cli') {
    $options = getopt("", ["gender:", "age:", "height:", "height-ft:", "height-in:", "weight:", "weight-lb:", "activity:", "goal:", "formula:", "output:"]);
    if (empty($options['gender']) || empty($options['age']) || empty($options['activity'])) {
        echo "Укажите --gender, --age, --activity\n";
        exit(1);
    }
    $gender = $options['gender'];
    $age = (float)$options['age'];
    $activity = $options['activity'];
    $goal = $options['goal'] ?? 'maintain';
    $formula = $options['formula'] ?? 'msj';
    $height = isset($options['height']) ? (float)$options['height'] : 0;
    $weight = isset($options['weight']) ? (float)$options['weight'] : 0;
    if ($height == 0 && isset($options['height-ft'])) {
        $ft = (int)$options['height-ft'];
        $inc = isset($options['height-in']) ? (int)$options['height-in'] : 0;
        $height = $ft * 30.48 + $inc * 2.54;
    }
    if ($weight == 0 && isset($options['weight-lb'])) {
        $weight = (float)$options['weight-lb'] * 0.453592;
    }
    if ($height <= 0 || $weight <= 0) {
        echo "Укажите рост и вес (метрические или имперские)\n";
        exit(1);
    }
    $params = ['gender' => $gender, 'age' => $age, 'height' => $height, 'weight' => $weight, 'activity' => $activity, 'goal' => $goal];
    $bmr = $formula == 'msj' ? bmrMifflin($gender, $age, $height, $weight) : bmrHarris($gender, $age, $height, $weight);
    $tdee = calculateTdee($bmr, $activity);
    $goalCal = calculateGoalCalories($tdee, $goal);
    $macros = calculateMacros($goalCal, $goal);
    $report = generateReport($params, $bmr, $tdee, $goalCal, $macros);
    echo $report . "\n";
    if (!empty($options['output'])) {
        file_put_contents($options['output'], $report);
        echo "Отчёт сохранён в {$options['output']}\n";
    }
    exit;
}

// ========== ВЕБ-ИНТЕРФЕЙС ==========
?>
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <title>Калькулятор калорий (PHP)</title>
    <style>
        body { font-family: 'Segoe UI', sans-serif; background: #f4f7fb; margin: 40px; }
        .container { max-width: 500px; margin: 0 auto; background: white; padding: 20px; border-radius: 16px; box-shadow: 0 2px 10px rgba(0,0,0,0.1); }
        label { display: inline-block; width: 100px; }
        input, select, button { margin: 8px 0; padding: 6px; }
        button { background: #3498db; color: white; border: none; padding: 8px 20px; border-radius: 4px; cursor: pointer; }
        .result { background: #e8f5e9; padding: 10px; border-radius: 8px; margin-top: 10px; white-space: pre-wrap; font-family: monospace; }
    </style>
</head>
<body>
<div class="container">
    <h1>🔥 Калькулятор калорий</h1>
    <form method="GET">
        <label>Пол:</label>
        <select name="gender">
            <option value="male" <?= isset($_GET['gender']) && $_GET['gender']=='male' ? 'selected' : '' ?>>Мужской</option>
            <option value="female" <?= isset($_GET['gender']) && $_GET['gender']=='female' ? 'selected' : '' ?>>Женский</option>
        </select><br>
        <label>Возраст (лет):</label>
        <input type="number" name="age" value="<?= isset($_GET['age']) ? $_GET['age'] : '' ?>" step="1"><br>
        <label>Рост (см):</label>
        <input type="number" name="height" value="<?= isset($_GET['height']) ? $_GET['height'] : '' ?>" step="0.1"><br>
        <label>Вес (кг):</label>
        <input type="number" name="weight" value="<?= isset($_GET['weight']) ? $_GET['weight'] : '' ?>" step="0.1"><br>
        <label>Активность:</label>
        <select name="activity">
            <option value="sedentary" <?= isset($_GET['activity']) && $_GET['activity']=='sedentary' ? 'selected' : '' ?>>Сидячий</option>
            <option value="light" <?= isset($_GET['activity']) && $_GET['activity']=='light' ? 'selected' : '' ?>>Лёгкая</option>
            <option value="moderate" <?= isset($_GET['activity']) && $_GET['activity']=='moderate' ? 'selected' : '' ?>>Умеренная</option>
            <option value="active" <?= isset($_GET['activity']) && $_GET['activity']=='active' ? 'selected' : '' ?>>Высокая</option>
            <option value="very_active" <?= isset($_GET['activity']) && $_GET['activity']=='very_active' ? 'selected' : '' ?>>Очень высокая</option>
        </select><br>
        <label>Цель:</label>
        <select name="goal">
            <option value="maintain" <?= isset($_GET['goal']) && $_GET['goal']=='maintain' ? 'selected' : '' ?>>Поддержание</option>
            <option value="lose" <?= isset($_GET['goal']) && $_GET['goal']=='lose' ? 'selected' : '' ?>>Похудение</option>
            <option value="gain" <?= isset($_GET['goal']) && $_GET['goal']=='gain' ? 'selected' : '' ?>>Набор массы</option>
        </select><br>
        <label>Формула:</label>
        <select name="formula">
            <option value="msj" <?= isset($_GET['formula']) && $_GET['formula']=='msj' ? 'selected' : '' ?>>Mifflin-St Jeor</option>
            <option value="hb" <?= isset($_GET['formula']) && $_GET['formula']=='hb' ? 'selected' : '' ?>>Harris-Benedict</option>
        </select><br>
        <button type="submit">Рассчитать</button>
    </form>
    <?php if (isset($_GET['gender']) && isset($_GET['age']) && isset($_GET['height']) && isset($_GET['weight']) && isset($_GET['activity'])): 
        $gender = $_GET['gender'];
        $age = (float)$_GET['age'];
        $height = (float)$_GET['height'];
        $weight = (float)$_GET['weight'];
        $activity = $_GET['activity'];
        $goal = $_GET['goal'] ?? 'maintain';
        $formula = $_GET['formula'] ?? 'msj';
        $params = ['gender' => $gender, 'age' => $age, 'height' => $height, 'weight' => $weight, 'activity' => $activity, 'goal' => $goal];
        $bmr = $formula == 'msj' ? bmrMifflin($gender, $age, $height, $weight) : bmrHarris($gender, $age, $height, $weight);
        $tdee = calculateTdee($bmr, $activity);
        $goalCal = calculateGoalCalories($tdee, $goal);
        $macros = calculateMacros($goalCal, $goal);
        $report = generateReport($params, $bmr, $tdee, $goalCal, $macros);
    ?>
        <div class="result"><?= htmlspecialchars($report) ?></div>
    <?php endif; ?>
</div>
</body>
</html>

File: calorie_calc.rs
// calorie_calc.rs - Калькулятор калорий на Rust (CLI)
use clap::{Arg, App};
use std::collections::HashMap;
use std::fs::File;
use std::io::Write;

lazy_static::lazy_static! {
    static ref ACTIVITY_LEVELS: HashMap<&'static str, f64> = {
        let mut m = HashMap::new();
        m.insert("sedentary", 1.2);
        m.insert("light", 1.375);
        m.insert("moderate", 1.55);
        m.insert("active", 1.725);
        m.insert("very_active", 1.9);
        m
    };
    static ref ACTIVITY_NAMES: HashMap<&'static str, &'static str> = {
        let mut m = HashMap::new();
        m.insert("sedentary", "Сидячий образ жизни");
        m.insert("light", "Лёгкая активность (1-3 дня/нед)");
        m.insert("moderate", "Умеренная активность (3-5 дней/нед)");
        m.insert("active", "Высокая активность (6-7 дней/нед)");
        m.insert("very_active", "Очень высокая активность (тяжёлая работа/спорт)");
        m
    };
    static ref GOAL_NAMES: HashMap<&'static str, &'static str> = {
        let mut m = HashMap::new();
        m.insert("maintain", "Поддержание веса");
        m.insert("lose", "Похудение (дефицит 15%)");
        m.insert("gain", "Набор массы (профицит 15%)");
        m
    };
    static ref GENDER_NAMES: HashMap<&'static str, &'static str> = {
        let mut m = HashMap::new();
        m.insert("male", "Мужской");
        m.insert("female", "Женский");
        m
    };
}

fn bmr_mifflin(gender: &str, age: f64, height: f64, weight: f64) -> f64 {
    if gender == "male" {
        10.0 * weight + 6.25 * height - 5.0 * age + 5.0
    } else {
        10.0 * weight + 6.25 * height - 5.0 * age - 161.0
    }
}

fn bmr_harris(gender: &str, age: f64, height: f64, weight: f64) -> f64 {
    if gender == "male" {
        88.362 + 13.397 * weight + 4.799 * height - 5.677 * age
    } else {
        447.593 + 9.247 * weight + 3.098 * height - 4.330 * age
    }
}

fn calculate_tdee(bmr: f64, activity: &str) -> f64 {
    bmr * ACTIVITY_LEVELS[activity]
}

fn calculate_goal_calories(tdee: f64, goal: &str) -> f64 {
    match goal {
        "maintain" => tdee,
        "lose" => tdee * 0.85,
        "gain" => tdee * 1.15,
        _ => tdee,
    }
}

fn calculate_macros(calories: f64, goal: &str) -> (f64, f64, f64) {
    let (protein_ratio, fat_ratio, carb_ratio) = match goal {
        "lose" => (0.35, 0.25, 0.40),
        "gain" => (0.30, 0.20, 0.50),
        _ => (0.30, 0.25, 0.45),
    };
    let protein = (calories * protein_ratio) / 4.0;
    let fat = (calories * fat_ratio) / 9.0;
    let carbs = (calories * carb_ratio) / 4.0;
    (protein, fat, carbs)
}

fn generate_report(params: &HashMap<String, String>, bmr: f64, tdee: f64, goal_cal: f64, protein: f64, fat: f64, carbs: f64) -> String {
    let mut lines = Vec::new();
    lines.push("=".repeat(50));
    lines.push("📋 ОТЧЁТ ПО КАЛОРИЯМ".to_string());
    lines.push("=".repeat(50));
    lines.push(format!("Пол: {}", GENDER_NAMES[params["gender"].as_str()]));
    lines.push(format!("Возраст: {} лет", params["age"]));
    lines.push(format!("Рост: {} см", params["height"]));
    lines.push(format!("Вес: {} кг", params["weight"]));
    lines.push(format!("Уровень активности: {}", ACTIVITY_NAMES[params["activity"].as_str()]));
    lines.push(format!("Цель: {}", GOAL_NAMES[params["goal"].as_str()]));
    lines.push("-".repeat(50));
    lines.push(format!("BMR (базовый метаболизм): {:.1} ккал/день", bmr));
    lines.push(format!("TDEE (суточная норма): {:.1} ккал/день", tdee));
    lines.push(format!("Рекомендуемая калорийность для цели: {:.1} ккал/день", goal_cal));
    lines.push("-".repeat(50));
    lines.push("🍽️ РАСПРЕДЕЛЕНИЕ ПО МАКРОНУТРИЕНТАМ:".to_string());
    lines.push(format!("Белки: {:.1} г ({:.0}%)", protein, (protein*4.0/goal_cal)*100.0));
    lines.push(format!("Жиры: {:.1} г ({:.0}%)", fat, (fat*9.0/goal_cal)*100.0));
    lines.push(format!("Углеводы: {:.1} г ({:.0}%)", carbs, (carbs*4.0/goal_cal)*100.0));
    lines.push("=".repeat(50));
    lines.join("\n")
}

fn main() {
    let matches = App::new("Calorie Calculator")
        .arg(Arg::with_name("gender").short("g").long("gender").takes_value(true).required(true).possible_values(&["male", "female"]))
        .arg(Arg::with_name("age").short("a").long("age").takes_value(true).required(true))
        .arg(Arg::with_name("height").long("height").takes_value(true).help("Рост (см)"))
        .arg(Arg::with_name("height-ft").long("height-ft").takes_value(true).help("Рост (футы)"))
        .arg(Arg::with_name("height-in").long("height-in").takes_value(true).default_value("0").help("Рост (дюймы)"))
        .arg(Arg::with_name("weight").long("weight").takes_value(true).help("Вес (кг)"))
        .arg(Arg::with_name("weight-lb").long("weight-lb").takes_value(true).help("Вес (фунты)"))
        .arg(Arg::with_name("activity").short("l").long("activity").takes_value(true).required(true).possible_values(&["sedentary", "light", "moderate", "active", "very_active"]))
        .arg(Arg::with_name("goal").short("o").long("goal").takes_value(true).default_value("maintain").possible_values(&["maintain", "lose", "gain"]))
        .arg(Arg::with_name("formula").short("f").long("formula").takes_value(true).default_value("msj").possible_values(&["msj", "hb"]))
        .arg(Arg::with_name("output").short("p").long("output").takes_value(true).help("Сохранить отчёт в файл"))
        .get_matches();

    let gender = matches.value_of("gender").unwrap();
    let age: f64 = matches.value_of("age").unwrap().parse().unwrap();
    let activity = matches.value_of("activity").unwrap();
    let goal = matches.value_of("goal").unwrap();
    let formula = matches.value_of("formula").unwrap();

    let height = if let Some(h) = matches.value_of("height") {
        h.parse().unwrap()
    } else {
        let ft: f64 = matches.value_of("height-ft").unwrap().parse().unwrap();
        let inc: f64 = matches.value_of("height-in").unwrap().parse().unwrap();
        ft * 30.48 + inc * 2.54
    };
    let weight = if let Some(w) = matches.value_of("weight") {
        w.parse().unwrap()
    } else {
        let lb: f64 = matches.value_of("weight-lb").unwrap().parse().unwrap();
        lb * 0.453592
    };

    let bmr = if formula == "msj" {
        bmr_mifflin(gender, age, height, weight)
    } else {
        bmr_harris(gender, age, height, weight)
    };
    let tdee = calculate_tdee(bmr, activity);
    let goal_cal = calculate_goal_calories(tdee, goal);
    let (protein, fat, carbs) = calculate_macros(goal_cal, goal);

    let mut params = HashMap::new();
    params.insert("gender".to_string(), gender.to_string());
    params.insert("age".to_string(), format!("{:.0}", age));
    params.insert("height".to_string(), format!("{:.1}", height));
    params.insert("weight".to_string(), format!("{:.1}", weight));
    params.insert("activity".to_string(), activity.to_string());
    params.insert("goal".to_string(), goal.to_string());

    let report = generate_report(&params, bmr, tdee, goal_cal, protein, fat, carbs);
    println!("{}", report);

    if let Some(output) = matches.value_of("output") {
        let mut file = File::create(output).unwrap();
        file.write_all(report.as_bytes()).unwrap();
        println!("Отчёт сохранён в {}", output);
    }
}
